File: backend/src/main/java/com/bob/ecommerceangularapp/service/ContentService.java
package com.bob.ecommerceangularapp.service;

import com.bob.ecommerceangularapp.dao.FaqEntryRepository;
import com.bob.ecommerceangularapp.dao.SiteBannerRepository;
import com.bob.ecommerceangularapp.dto.FaqEntryRequest;
import com.bob.ecommerceangularapp.dto.SiteBannerRequest;
import com.bob.ecommerceangularapp.entity.FaqEntry;
import com.bob.ecommerceangularapp.entity.SiteBanner;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

/**
 * A deliberately small CMS (roadmap #17): a single site-wide announcement banner and a list of FAQ
 * entries, both previously hardcoded in the frontend, now editable by an admin without a deploy.
 */
@Service
public class ContentService {

    private final SiteBannerRepository siteBannerRepository;
    private final FaqEntryRepository faqEntryRepository;

    public ContentService(SiteBannerRepository siteBannerRepository, FaqEntryRepository faqEntryRepository) {
        this.siteBannerRepository = siteBannerRepository;
        this.faqEntryRepository = faqEntryRepository;
    }

    // ----- storefront -----

    /** The banner, if one is configured and turned on. */
    @Transactional(readOnly = true)
    public Optional<SiteBanner> getActiveBanner() {
        return currentBanner().filter(SiteBanner::isActive);
    }

    @Transactional(readOnly = true)
    public List<FaqEntry> listActiveFaq() {
        return faqEntryRepository.findByActiveTrueOrderBySortOrderAscIdAsc();
    }

    // ----- admin -----

    /** The banner row for editing, even if currently turned off; a blank shell if none exists yet. */
    @Transactional(readOnly = true)
    public SiteBanner getBannerForAdmin() {
        return currentBanner().orElseGet(SiteBanner::new);
    }

    @Transactional
    public SiteBanner saveBanner(SiteBannerRequest request) {
        SiteBanner banner = currentBanner().orElseGet(SiteBanner::new);
        banner.setMessage(request.message().trim());
        banner.setLinkUrl(blankToNull(request.linkUrl()));
        banner.setLinkText(blankToNull(request.linkText()));
        banner.setActive(request.active());
        return siteBannerRepository.save(banner);
    }

    @Transactional(readOnly = true)
    public List<FaqEntry> listFaqForAdmin() {
        return faqEntryRepository.findAllByOrderBySortOrderAscIdAsc();
    }

    @Transactional
    public FaqEntry saveFaq(FaqEntryRequest request) {
        FaqEntry entry = request.id() != null
                ? faqEntryRepository.findById(request.id())
                        .orElseThrow(() -> new IllegalArgumentException("FAQ entry not found: " + request.id()))
                : new FaqEntry();
        entry.setQuestion(request.question().trim());
        entry.setAnswer(request.answer().trim());
        entry.setSortOrder(request.sortOrder());
        entry.setActive(request.active());
        return faqEntryRepository.save(entry);
    }

    @Transactional
    public void deleteFaq(Long id) {
        if (!faqEntryRepository.existsById(id)) {
            throw new IllegalArgumentException("FAQ entry not found: " + id);
        }
        faqEntryRepository.deleteById(id);
    }

    private Optional<SiteBanner> currentBanner() {
        return siteBannerRepository.findAll().stream().findFirst();
    }

    private static String blankToNull(String s) {
        return (s == null || s.isBlank()) ? null : s.trim();
    }
}

File: frontend/angular-ecommerce/src/app/services/content.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable, of } from 'rxjs';
import { catchError } from 'rxjs/operators';

import { environment } from '../../environments/environment';

export interface SiteBanner {
  id: number;
  message: string;
  linkUrl?: string | null;
  linkText?: string | null;
  active: boolean;
}

export interface FaqEntry {
  id: number;
  question: string;
  answer: string;
  sortOrder: number;
  active: boolean;
}

/** Public reads for the CMS-managed content (roadmap #17): the site banner and the FAQ list. */
@Injectable({ providedIn: 'root' })
export class ContentService {

  private readonly baseUrl = `${environment.apiUrl}/content`;

  constructor(private http: HttpClient) {}

  /** Null when no banner is configured or it's turned off (the server returns 204 for both). */
  getBanner(): Observable<SiteBanner | null> {
    return this.http.get<SiteBanner>(`${this.baseUrl}/banner`).pipe(catchError(() => of(null)));
  }

  getFaq(): Observable<FaqEntry[]> {
    return this.http.get<FaqEntry[]>(`${this.baseUrl}/faq`).pipe(catchError(() => of([])));
  }
}
